### ISSUE TITLE
Error when building both static and shared libs on Windows with MSVC

### ISSUE THREAD

**Issue by adriendelsalle** [LLNL/sundials#454]
Description
---

It's currently not possible to build both static and shared versions of libraries on Windows when using the MSVC-based toolchain because both variants produce a `.lib` file with the exact same output name.

For the shared libs it represents the export file and for static libs, the library itself. 
It would be nice to detect such cases and automatically suffix the output name of the static lib with `_static` to disambiguate. It is also well documented in a [`conda-forge` enhancement proposal](https://github.com/conda-forge/cfep/blob/main/cfep-18.md#windows).

I'll open a PR to discuss a possible fix!

Steps to reproduce
---

I'm building the project using a conda environment:
```
micromamba create -n sundials -y c-compiler cmake ninja
micromamba activate sundials
cmake -LAH -G "Ninja" -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DBUILD_STATIC_LIBS=ON -S .
ninja -C build
```
Then you get at the last step:
```
ninja: error: build.ninja:997: multiple rules generate bin/sundials_core.lib
```

**Comment by balos1** [comment]
Hi @adriendelsalle, yes this is a known issue. A fix is currently in development. We would appreciate it if you could test the fix out when it is ready.

**Comment by adriendelsalle** [comment]
Sure! Could you share a link when ready for testing? Thanks for the quick reply :)

**Comment by adriendelsalle** [comment]
I also noticed that the `SUNDIALS_EXPORT` macro seems invalid when building both variants:

In the following code, `sundials_core` refers to the shared lib when building both static and shared ones;
https://github.com/LLNL/sundials/blob/b685654c6958699821fc68a822b66076fec8cae7/src/sundials/CMakeLists.txt#L159-L164

causing `CMake`'s `generate_export_header` to generate the following `sundials_export.h`:
```
#ifdef SUNDIALS_STATIC_DEFINE
#  define SUNDIALS_EXPORT
#  define SUNDIALS_NO_EXPORT
#else
#  ifndef SUNDIALS_EXPORT
#    ifdef sundials_core_EXPORTS
        /* We are building this library */
#      define SUNDIALS_EXPORT __declspec(dllexport)
#    else
        /* We are using this library */
#      define SUNDIALS_EXPORT __declspec(dllimport)
#    endif
#  endif
```
but `SUNDIALS_STATIC_DEFINE` is set as `PRIVATE` compile definition of static targets:
https://github.com/LLNL/sundials/blob/b685654c6958699821fc68a822b66076fec8cae7/cmake/macros/SundialsAddLibrary.cmake#L203
and
https://github.com/LLNL/sundials/blob/b685654c6958699821fc68a822b66076fec8cae7/cmake/macros/SundialsAddLibrary.cmake#L299

Causing the `SUNDIALS_EXPORT` macro to expand to `__declspec(dllimport)` even for the static target. Using MSVC it leads to this kind of errors: ` error LNK2019: unresolved external symbol __imp_CVodeSetLinearSolver` (see https://stackoverflow.com/a/5159395)


I think `SUNDIALS_STATIC_DEFINE` should be a `PUBLIC` compile definition for static libs

**Comment by balos1** [comment]
Closed by #455.
